### PR TITLE
Graduate EndpointSlice API to GA

### PR DESCRIPTION
**Pull request by swetharepakula**
* removes discovery v1alpha1 API
  * removes EndpointSlice Topology field from GA API
  * adds per Endpoint Zone field in GA API

#### What type of PR is this?
/kind feature
/kind api-change

#### What this PR does / why we need it:
Graduates the EndpointSlice API to GA and removes the v1alpha1 API.

#### Does this PR introduce a user-facing change?

```release-note
EndpointSlice API is now GA. The EndpointSlice topology field has been removed from the GA API and will be replaced by a new per Endpoint Zone field. If the topology field was previously used, it will be converted into an annotation in the v1 Resource. The discovery.k8s.io/v1alpha1 API is removed.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

- [KEP]: [EndpointSlice Enhancement](https://github.com/kubernetes/enhancements/blob/8413469b89851d094eb22813a06594bd4a6d36a4/keps/sig-network/0752-endpointslices/README.md)

/cc @aojea @andrewsykim @liggitt @robscott @wojtek-t 
/assign @thockin

### CODE REVIEW

**Comment by robscott** [comment]
Thanks for the work on this!

/ok-to-test
/sig network
/triage accepted
/priority important-soon

File: staging/src/k8s.io/api/discovery/v1/well_known_labels.go [hunk 1]
```diff
@@ -25,4 +25,8 @@ const (
 	// same cluster. It is highly recommended to configure this label for all
 	// EndpointSlices.
 	LabelManagedBy = "endpointslice.kubernetes.io/managed-by"
+	// LabelSkipMirror can be set to true on an Endpoints resource to indicate
+	// that the EndpointSliceMirroring controller should not mirror this
+	// resource with EndpointSlices.
+	LabelSkipMirror = "endpointslice.kubernetes.io/skip-mirror"
```
Comment (thockin) [inline staging/src/k8s.io/api/discovery/v1/well_known_labels.go:31]: Why is this in here?  Should it be a different commit?

File: staging/src/k8s.io/kubectl/go.mod [hunk 2]
```diff
@@ -13,10 +13,9 @@ require (
 	github.com/evanphx/json-patch v4.9.0+incompatible
 	github.com/exponent-io/jsonpath v0.0.0-20151013193312-d6023ce2651d
 	github.com/fatih/camelcase v1.0.0
-	github.com/fsnotify/fsnotify v1.4.9 // indirect
 	github.com/fvbommel/sortorder v1.0.1
 	github.com/go-openapi/spec v0.19.3
-	github.com/golangplus/testing v0.0.0-20180327235837-af21d9c3145e // indirect
```
Comment (thockin) [inline staging/src/k8s.io/kubectl/go.mod:19]: I'd like to underatand these changes too

File: staging/src/k8s.io/kubectl/go.sum [hunk 3]
```diff
@@ -155,8 +155,13 @@ github.com/golang/protobuf v1.4.1/go.mod h1:U8fpvMrcmy5pZrNK1lt4xCsGvpyWQ/VVv6QD
 github.com/golang/protobuf v1.4.2/go.mod h1:oDoupMAO8OvCJWAcko0GGGIgR6R6ocIYbsSw735rRwI=
 github.com/golang/protobuf v1.4.3 h1:JjCZWpVbqXDqFVmTfYWEVTMIYrL/NPdPSCHPJ0T/raM=
 github.com/golang/protobuf v1.4.3/go.mod h1:oDoupMAO8OvCJWAcko0GGGIgR6R6ocIYbsSw735rRwI=
-github.com/golangplus/testing v0.0.0-20180327235837-af21d9c3145e h1:KhcknUwkWHKZPbFy2P7jH5LKJ3La+0ZeknkkmrSgqb0=
-github.com/golangplus/testing v0.0.0-20180327235837-af21d9c3145e/go.mod h1:0AA//k/eakGydO4jKRoRL2j92ZKSzTgj9tclaCrvXHk=
+github.com/golangplus/bytes v0.0.0-20160111154220-45c989fe5450/go.mod h1:Bk6SMAONeMXrxql8uvOKuAZSu8aM5RUGv+1C6IJaEho=
```
Comment (thockin) [inline staging/src/k8s.io/kubectl/go.sum:158]: Same with other mod changes - can we explain?

File: go.mod [hunk 4]
```diff
@@ -26,6 +26,7 @@ require (
 	github.com/boltdb/bolt v1.3.1 // indirect
 	github.com/caddyserver/caddy v1.0.3
 	github.com/clusterhq/flocker-go v0.0.0-20160920122132-2b8b7259d313
+	github.com/codegangsta/negroni v1.0.0 // indirect
```
[the remark this answers is not in the document]
Comment (liggitt) [reply go.mod:29]: `master` just switched to go1.16 which trims out a bit more... this was likely run locally with go.15.x and tried to put it back in

File: staging/src/k8s.io/api/discovery/v1/types.go [hunk 5]
```diff
@@ -165,12 +155,15 @@ type EndpointPort struct {
 	// This field follows standard Kubernetes label syntax.
 	// Un-prefixed names are reserved for IANA standard service names (as per
 	// RFC-6335 and http://www.iana.org/assignments/service-names).
-	// Non-standard protocols should use prefixed names.
-	// Default is empty string.
+	// Non-standard protocols should use prefixed names such as
+	// mycompany.com/my-custom-protocol.
```
[the remark this answers is not in the document]
Comment (robscott) [reply staging/src/k8s.io/api/discovery/v1/types.go:159]: I  think this is something I missed porting back to v1alpha1 but matches what we [already have in v1beta1](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/api/discovery/v1beta1/types.go#L164-L165). The git diff is weird here because v1alpha1 is being removed at the same time v1 is being added, so it's treating everything as a rename (and catching changes I or others missed adding to v1alpha1).

File: staging/src/k8s.io/api/discovery/v1/well_known_labels.go (hunk 1 above)
Comment (robscott) [reply staging/src/k8s.io/api/discovery/v1/well_known_labels.go:31]: Also something that made it to [v1beta1](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/api/discovery/v1beta1/well_known_labels.go#L31) but not v1alpha1.

File: staging/src/k8s.io/api/discovery/v1/types.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (liggitt) [reply staging/src/k8s.io/api/discovery/v1/types.go:159]: I would suggest removing v1alpha1 in a separate commit (or maybe even a separate PR)

File: go.mod (hunk 4 above)
[the remark this answers is not in the document]
Comment (swetharepakula) [reply go.mod]: The mod changes were due to the go version mismatch. Once I updated to go1.16 the go.mod changes were all gone. Thanks @liggitt for point that out!

File: staging/src/k8s.io/kubectl/go.mod (hunk 2 above)
Comment (swetharepakula) [reply staging/src/k8s.io/kubectl/go.mod]: go.mod changes were do to the go version mismatch. Switching to go1.16 removed these changes.

File: staging/src/k8s.io/kubectl/go.sum (hunk 3 above)
Comment (swetharepakula) [reply staging/src/k8s.io/kubectl/go.sum]: go.mod changes were do to the go version mismatch. Switching to go1.16 removed these changes.

File: pkg/apis/discovery/v1beta1/conversion.go [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+				valuesSize += len(v)
+			}
+			valueCounts[v]++
+		}
+	}
+
+	approximateSize := approximateAnnotationSize(keyMap, valueCounts, len(in.Endpoints), keysSize, valuesSize)
+
+	maxSize := 256*(1<<10) - sizeOfAnnotations(out.Annotations)
+
```
Comment (liggitt) [inline pkg/apis/discovery/v1beta1/conversion.go:75]: I haven't reviewed this in detail, but roundtripping API fields to json-serialized annotations in other versions has a lot of sharp edges. Some scenarios HPA hit when it took a similar path (serializing fields from an earlier API version into annotations) that you should be aware of and test for:

* Make sure the value you're marshaling/unmarshaling into the annotation is a stable API structure (HPA marshaled an internal API struct accidentally)
* Always control the annotation value for this key (by clearing it or setting it) when converting to v1
* Always drop the annotation value for this key when converting to/from v1beta1 (otherwise we could break the ability to read persisted data from etcd)

xref https://github.com/kubernetes/kubernetes/pull/89963, https://github.com/kubernetes/kubernetes/pull/89963, and https://github.com/kubernetes/kubernetes/issues/89964

File: pkg/apis/discovery/v1beta1/conversion.go [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+		return err
+	}
+
+	annotation, ok := in.Annotations[v1beta1Topology]
+	if !ok {
+		return nil
+	}
+
+	var epsTop EndpointSliceTopology
+	if err := json.Unmarshal([]byte(annotation), &epsTop); err != nil {
```
Comment (liggitt) [inline pkg/apis/discovery/v1beta1/conversion.go:167]: since this is a pretty high volume / high write rate API, are we concerned about the impact of this?

File: pkg/apis/discovery/v1beta1/conversion.go [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+				stringIndex[k] = kIndex
+			}
+
+			if vIndex, ok = stringIndex[v]; !ok {
+				vIndex = len(strings)
+				strings = append(strings, v)
+			}
+			topology[kIndex] = vIndex
+		}
+		topologies = append(topologies, topology)
```
Comment (liggitt) [inline pkg/apis/discovery/v1beta1/conversion.go:133]: This results in a v1 object that has a parallel slice of endpoints and topology annotations. Does that means if a v1 API client adds or removes an endpoint from the list, it is also responsible for rewriting the annotation to keep it in sync? Are we confident clients will be able to do that correctly? What are the implications if they do not?

File: staging/src/k8s.io/api/discovery/v1/types.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (thockin) [reply staging/src/k8s.io/api/discovery/v1/types.go:159]: changing v1a1 -> v1 made this PR easy to review

File: pkg/apis/discovery/v1beta1/conversion.go (hunk 6 above)
Comment (thockin) [reply pkg/apis/discovery/v1beta1/conversion.go:75]: > Always control the annotation value for this key (by clearing it or setting it) when converting to v1
> Always drop the annotation value for this key when converting to/from v1beta1

Good points.  I thought there might be some trick to this, but it seems straightforward.  Ongoing discussion IRL - maybe we won't do annnotation.  More as we figure it out.

File: pkg/apis/discovery/v1beta1/conversion.go (hunk 7 above)
Comment (thockin) [reply pkg/apis/discovery/v1beta1/conversion.go:167]: use json-iterator?  Or keep it a field, as you suggested elsewhere

File: pkg/apis/discovery/v1beta1/conversion.go (hunk 8 above)
Comment (thockin) [reply pkg/apis/discovery/v1beta1/conversion.go:133]: THis was the plan but a field is looking better and better

File: pkg/apis/discovery/validation/validation.go [hunk 9]
```diff
@@ -38,10 +37,9 @@ var (
 		string(api.ProtocolUDP),
 		string(api.ProtocolSCTP),
 	)
-	maxTopologyLabels = 16
-	maxAddresses      = 100
-	maxPorts          = 100
-	maxEndpoints      = 1000
+	maxAddresses = 100
+	maxPorts     = 100
```
Comment (aojea) [inline pkg/apis/discovery/validation/validation.go:41]: this is going to collide with https://github.com/kubernetes/kubernetes/pull/99795 :)

File: pkg/apis/discovery/v1beta1/conversion.go [hunk 10]
```diff
@@ -0,0 +1,239 @@
...
+
+	digitsPerKey := len(strconv.Itoa(numKeys))
+	digitsPerValue := len(strconv.Itoa(numValues))
+
+	// "endpointslice.kubernetes.io/v1beta1-topology" 44 bytes
+	// {"s":[],"t":[]} -> 15 bytes
+	// each topology needs {} -> 3 bytes
+	// each topology "":"", + keyDigits + valueDigits per key -> 6 + keyDigits + valueDigits
+
+	return 44 + 16 + 3*(numKeys+numValues) + numKeyBytes + numValueBytes + numEndpoints*(3+numKeys*(6+digitsPerKey+digitsPerValue))
```
Comment (aojea) [inline pkg/apis/discovery/v1beta1/conversion.go:238]: the comment says 15 bytes and you have a 16, is that correct?

**Comment by swetharepakula** [comment]
After some off line discussion, we are going to change the approach to this. Instead of having an annotation, we are going to be keeping a `DeprecatedTopology` field on the v1 API that will be removed when v1beta1 is removed. We plan on making that field read only in the V1 API. This should greatly reduce the complexity of the conversion and make it clearer to users of the topology field how to proceed.

**Comment by swetharepakula** [comment]
I have made an update to PR with the new approach outlined in the above [comment](https://github.com/kubernetes/kubernetes/pull/99662#issuecomment-791033604). It is ready for review again. Thank you all for all the feedback and help!

File: pkg/apis/discovery/validation/validation.go [hunk 11]
```diff
@@ -110,11 +110,11 @@ func validateEndpoints(endpoints []discovery.Endpoint, addrType discovery.Addres
...
 
-		topologyPath := idxPath.Child("topology")
-		if len(endpoint.Topology) > maxTopologyLabels {
-			allErrs = append(allErrs, field.TooMany(topologyPath, len(endpoint.Topology), maxTopologyLabels))
+		topologyPath := idxPath.Child("deprecatedTopology")
+		if len(endpoint.DeprecatedTopology) > maxTopologyLabels {
+			allErrs = append(allErrs, field.TooMany(topologyPath, len(endpoint.DeprecatedTopology), maxTopologyLabels))
 		}
-		allErrs = append(allErrs, metavalidation.ValidateLabels(endpoint.Topology, topologyPath)...)
+		allErrs = append(allErrs, metavalidation.ValidateLabels(endpoint.DeprecatedTopology, topologyPath)...)
```
Comment (thockin) [inline pkg/apis/discovery/validation/validation.go:117]: In my sketch I had

```
+               if _, found := endpoint.DeprecatedTopology[v1.LabelTopologyZone]; found {
+                       allErrs = append(allErrs, field.InternalError(topologyPath.Key(v1.LabelTopologyZone), fmt.Errorf("reserved key was not removed in conversion")))
+               }
```

Do you think that's a bad idea?

File: pkg/registry/discovery/endpointslice/strategy_test.go [hunk 12]
```diff
@@ -628,3 +630,324 @@ func TestPrepareForUpdate(t *testing.T) {
...
+					{
+						NodeName:           utilpointer.StringPtr("node-1"),
+						DeprecatedTopology: map[string]string{"key": "value"},
+					},
+				},
+			},
+		},
+	}
+
+	for _, tc := range testcases {
```
Comment (thockin) [inline pkg/registry/discovery/endpointslice/strategy_test.go:938]: Great testcases

**Review by thockin — COMMENTED** [review]
This looks pretty solid to me

File: pkg/apis/discovery/validation/validation.go (hunk 11 above)
Comment (thockin) [reply pkg/apis/discovery/validation/validation.go:117]: This is the only comment to resolve still.

Did you figure out why the v1b1 controller, setting region, wasn't coming thru to v1 deprecatedTopology?  Highly likely it was PEBKAC, but would be good to verify

File: pkg/apis/discovery/install/install.go [hunk 13]
```diff
@@ -34,7 +34,7 @@ func init() {
 // Install registers the API group and adds types to a scheme
 func Install(scheme *runtime.Scheme) {
 	utilruntime.Must(discovery.AddToScheme(scheme))
-	utilruntime.Must(v1alpha1.AddToScheme(scheme))
 	utilruntime.Must(v1beta1.AddToScheme(scheme))
-	utilruntime.Must(scheme.SetVersionPriority(v1beta1.SchemeGroupVersion, v1alpha1.SchemeGroupVersion))
+	utilruntime.Must(v1.AddToScheme(scheme))
+	utilruntime.Must(scheme.SetVersionPriority(v1.SchemeGroupVersion, v1beta1.SchemeGroupVersion))
```
Comment (liggitt) [inline pkg/apis/discovery/install/install.go:39]: prefer v1beta1 over v1 until 1.22, so that etcd storage writes using v1beta1, so that 1.20 API servers can read the objects we write.

File: test/integration/etcd/data.go [hunk 14]
```diff
@@ -191,10 +191,18 @@ func GetEtcdStorageDataForNamespace(namespace string) map[schema.GroupVersionRes
...
+			Stub:             `{"metadata": {"name": "slicev1"}, "addressType": "IPv4", "protocol": "TCP", "ports": [], "endpoints": []}`,
+			ExpectedEtcdPath: "/registry/endpointslices/" + namespace + "/slicev1",
+		},
+		// --
+
 		// k8s.io/kubernetes/pkg/apis/discovery/v1beta1
 		gvr("discovery.k8s.io", "v1beta1", "endpointslices"): {
 			Stub:             `{"metadata": {"name": "slicev1beta1"}, "addressType": "IPv4", "protocol": "TCP", "ports": [], "endpoints": []}`,
 			ExpectedEtcdPath: "/registry/endpointslices/" + namespace + "/slicev1beta1",
+			ExpectedGVK:      gvkP("discovery.k8s.io", "v1", "EndpointSlice"),
```
Comment (liggitt) [inline test/integration/etcd/data.go:205]: storage version should remain v1beta1 until 1.22 (see comment above in install about preferring v1beta1 until then)

File: pkg/apis/discovery/types.go [hunk 15]
```diff
@@ -87,26 +87,19 @@ type Endpoint struct {
...
-	// * kubernetes.io/hostname: the value indicates the hostname of the node
-	//   where the endpoint is located. This should match the corresponding
-	//   node label.
-	// * topology.kubernetes.io/zone: the value indicates the zone where the
-	//   endpoint is located. This should match the corresponding node label.
-	// * topology.kubernetes.io/region: the value indicates the region where the
-	//   endpoint is located. This should match the corresponding node label.
-	// This field is deprecated and will be removed in future api versions.
+	// deprecatedTopology is deprecated and only retained for round-trip
+	// compatbility with v1beta1 Topology field.  When v1beta1 is removed, this
```
Comment (liggitt) [inline pkg/apis/discovery/types.go:91]: nit:
```suggestion
	// compatibility with v1beta1 Topology field.  When v1beta1 is removed, this
```

File: pkg/controlplane/storageversionhashdata/data.go [hunk 16]
```diff
@@ -66,7 +66,8 @@ var GVRToStorageVersionHash = map[string]string{
 	"certificates.k8s.io/v1beta1/certificatesigningrequests":               "95fRKMXA+00=",
 	"coordination.k8s.io/v1beta1/leases":                                   "gqkMMb/YqFM=",
 	"coordination.k8s.io/v1/leases":                                        "gqkMMb/YqFM=",
-	"discovery.k8s.io/v1beta1/endpointslices":                              "Nx3SIv6I0mE=",
+	"discovery.k8s.io/v1/endpointslices":                                   "qgS0xkrxYAI=",
+	"discovery.k8s.io/v1beta1/endpointslices":                              "qgS0xkrxYAI=",
```
Comment (liggitt) [inline pkg/controlplane/storageversionhashdata/data.go:70]: storage version hash shouldn't change until 1.22

File: pkg/registry/discovery/endpointslice/strategy.go [hunk 17]
```diff
@@ -139,3 +145,37 @@ func dropDisabledFieldsOnUpdate(oldEPS, newEPS *discovery.EndpointSlice) {
...
+// dropTopologyOnV1 on V1 request wipes the DeprecatedTopology field  and copies
+// the NodeName value into DeprecatedTopology
+func dropTopologyOnV1(ctx context.Context, oldEPS, newEPS *discovery.EndpointSlice) {
+	if info, ok := genericapirequest.RequestInfoFrom(ctx); ok {
+
+		// endpointListUpdated should be true for creates
+		endpointListUpdated := true
+		if oldEPS != nil {
+			endpointListUpdated = !apiequality.Semantic.DeepEqual(oldEPS.Endpoints, newEPS.Endpoints)
+		}
```
Comment (liggitt) [inline pkg/registry/discovery/endpointslice/strategy.go:158]: nit: move this inside the version check so we don't do work we won't use on v1beta1 requests

File: pkg/registry/discovery/endpointslice/strategy.go [hunk 18]
```diff
@@ -139,3 +145,37 @@ func dropDisabledFieldsOnUpdate(oldEPS, newEPS *discovery.EndpointSlice) {
...
+
+		requestGV := schema.GroupVersion{Group: info.APIGroup, Version: info.APIVersion}
+		if requestGV == discoveryv1.SchemeGroupVersion {
```
Comment (liggitt) [inline pkg/registry/discovery/endpointslice/strategy.go:161]: suggest checking our version and returning early for v1beta1 specifically. we'd want to drop deprecated topology the same way in future versions without needing to revisit this code.

```
func dropTopologyOnV1(ctx context.Context, oldEPS, newEPS *discovery.EndpointSlice) {
  if info, ok := genericapirequest.RequestInfoFrom(ctx); ok {
    // ... return early if v1beta1
  }
  ... check deepequal, drop, etc...
```

File: pkg/registry/discovery/endpointslice/strategy.go [hunk 19]
```diff
@@ -139,3 +145,37 @@ func dropDisabledFieldsOnUpdate(oldEPS, newEPS *discovery.EndpointSlice) {
...
+					ep.DeprecatedTopology = nil
+				}
+
+				// Set topology[hostname] to match NodeName for compatibility with old clients
+				if ep.NodeName != nil {
+					if ep.DeprecatedTopology == nil {
+						ep.DeprecatedTopology = make(map[string]string, 1)
+					}
+					ep.DeprecatedTopology[corev1.LabelHostname] = *ep.NodeName
+				}
```
Comment (liggitt) [inline pkg/registry/discovery/endpointslice/strategy.go:177]: why isn't this limited to `endpointListUpdated` as well? this means a v1 writer touching only metadata fields (like the garbage collector or something) would cause nodeName to get propagated into deprecatedTopology. is that what we want?

File: pkg/apis/discovery/v1beta1/conversion.go [hunk 20]
```diff
@@ -0,0 +1,74 @@
...
+	if zone, ok := in.Topology[corev1.LabelTopologyZone]; ok {
+		out.Zone = &zone
+		delete(out.DeprecatedTopology, corev1.LabelTopologyZone)
+	}
+
+	// If zone was the only field in the map or topology was empty set DeprecatedTopology
+	// to nil
+	if len(out.DeprecatedTopology) == 0 {
+		out.DeprecatedTopology = nil
+	}
```
Comment (liggitt) [inline pkg/apis/discovery/v1beta1/conversion.go:47]: put all of this in an `if in.Topology != nil` check so we don't allocate maps we won't use

File: pkg/apis/discovery/v1beta1/conversion.go [hunk 21]
```diff
@@ -0,0 +1,74 @@
...
+
+	return nil
+}
+
+func Convert_discovery_Endpoint_To_v1beta1_Endpoint(in *discovery.Endpoint, out *v1beta1.Endpoint, s conversion.Scope) error {
```
Comment (liggitt) [inline pkg/apis/discovery/v1beta1/conversion.go:52]: I thought when converting to v1, we were going to drop the hostname label from DeprecatedTopology if it matched nodeName, so under typical circumstances, a v1 client would never see the deprecatedTopology field at all

File: pkg/apis/discovery/install/install.go (hunk 13 above)
Comment (thockin) [reply pkg/apis/discovery/install/install.go:39]: good catch

File: pkg/registry/discovery/endpointslice/strategy.go (hunk 18 above)
Comment (thockin) [reply pkg/registry/discovery/endpointslice/strategy.go:161]: that has a nice property of forcing us to look at this when v1b1 goes away

File: pkg/apis/discovery/v1beta1/conversion.go (hunk 21 above)
Comment (thockin) [reply pkg/apis/discovery/v1beta1/conversion.go:52]: I think we agreed to leave it alone - if the user wrote 2 different values they get to keep them, unless they write thru v1.  I don't see any real reason to unpack it further?

**Comment by liggitt** [comment]
I agree this looks close. The only open question I have is whether we want `deprecatedTopology[hostname]` served in v1 when it matches nodeName.

File: pkg/apis/discovery/v1beta1/conversion.go (hunk 21 above)
Comment (liggitt) [reply pkg/apis/discovery/v1beta1/conversion.go:52]: the propagation of nodeName → deprecatedTopology[hostname] when writing via v1 is what I'm trying to think through... won't that make `deprecatedTopology[hostname]` always be populated in v1? I thought we wanted that empty/nil/omitted in the happy "write via v1 and then read via v1" path

Comment (thockin) [reply pkg/apis/discovery/v1beta1/conversion.go:52]: You're right that `deprecatedTopology[hostname]` should show up in v1 but it doesn't (in my branch).  Hmm.  Will have to try this branch tomorrow.

File: pkg/registry/discovery/endpointslice/strategy.go (hunk 19 above)
Comment (swetharepakula) [reply pkg/registry/discovery/endpointslice/strategy.go:177]: I think we need to have nodeName propagated in deprecatedTopology, otherwise older kube-proxies would not be able to read nodeName information since we don't do anything special in the conversion in regards to the hostname topology label. When we remove v1beta1, we could then remove this logic to stop populating topology and eventually deprecatedTopology will be an unused field.

Comment (liggitt) [reply pkg/registry/discovery/endpointslice/strategy.go:177]: > I think we need to have nodeName propagated in deprecatedTopology, otherwise older kube-proxies would not be able to read nodeName information since we don't do anything special in the conversion in regards to the hostname topology label.

When endpoints are being modified via v1, I agree. If the endpoint was written via v1beta1, doing this unconditionally (not just when endpoints were modified via v1) means we'll copy nodeName into topology

Comment (swetharepakula) [reply pkg/registry/discovery/endpointslice/strategy.go:177]: We still have the version check, but I see your point that it makes sense only to do this when the v1 client is writing to Endpoints

File: pkg/apis/discovery/v1beta1/conversion.go (hunk 21 above)
Comment (thockin) [reply pkg/apis/discovery/v1beta1/conversion.go:52]: OK, I confirm that this DOES set `deprecatedTopology[hostname]`, as expected.  The question is whether we care?

I'll see what a quick fix woould look like

Comment (thockin) [reply pkg/apis/discovery/v1beta1/conversion.go:52]: This seems to work - Swetha, what do you think?  I did not update tests, but manual testing seems to work.   I think this solves the other issue too?

```
diff --git a/pkg/apis/discovery/v1beta1/conversion.go b/pkg/apis/discovery/v1beta1/conversion.go
index f741c5955c8..507a28afcbe 100644
--- a/pkg/apis/discovery/v1beta1/conversion.go
+++ b/pkg/apis/discovery/v1beta1/conversion.go
@@ -40,6 +40,16 @@ func Convert_v1beta1_Endpoint_To_discovery_Endpoint(in *v1beta1.Endpoint, out *d
 		delete(out.DeprecatedTopology, corev1.LabelTopologyZone)
 	}
 
+	// Remove hostname from the topology map ONLY IF it is the same value as
+	// nodeName.  This preserves the (rather odd) ability to have different
+	// values for topology[hostname] and nodename in v1beta1, without showing
+	// duplicate values in v1.
+	if node, ok := in.Topology[corev1.LabelHostname]; ok {
+		if out.NodeName != nil && node == *out.NodeName {
+			delete(out.DeprecatedTopology, corev1.LabelHostname)
+		}
+	}
+
 	// If zone was the only field in the map or topology was empty set DeprecatedTopology
 	// to nil
 	if len(out.DeprecatedTopology) == 0 {
@@ -55,7 +65,7 @@ func Convert_discovery_Endpoint_To_v1beta1_Endpoint(in *discovery.Endpoint, out
 	}
 
 	// If no deprecated topology or zone field, no conversion is necessary
-	if in.DeprecatedTopology == nil && in.Zone == nil {
+	if in.DeprecatedTopology == nil && in.Zone == nil && in.NodeName == nil {
 		return nil
 	}
 
@@ -70,5 +80,12 @@ func Convert_discovery_Endpoint_To_v1beta1_Endpoint(in *discovery.Endpoint, out
 		out.Topology[corev1.LabelTopologyZone] = *in.Zone
 	}
 
+	// Add hostname into the topology map ONLY IF it is not already present.
+	// This preserves the (rather odd) ability to have different values for
+	// topology[hostname] and nodename in v1beta1.
+	if in.NodeName != nil && out.Topology[corev1.LabelHostname] == "" {
+		out.Topology[corev1.LabelHostname] = *in.NodeName
+	}
+
 	return nil
 }
diff --git a/pkg/registry/discovery/endpointslice/strategy.go b/pkg/registry/discovery/endpointslice/strategy.go
index d8be219fd07..bf287afbf2a 100644
--- a/pkg/registry/discovery/endpointslice/strategy.go
+++ b/pkg/registry/discovery/endpointslice/strategy.go
@@ -19,7 +19,6 @@ package endpointslice
 import (
 	"context"
 
-	corev1 "k8s.io/api/core/v1"
 	discoveryv1 "k8s.io/api/discovery/v1"
 	apiequality "k8s.io/apimachinery/pkg/api/equality"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
@@ -146,8 +145,7 @@ func dropDisabledFieldsOnUpdate(oldEPS, newEPS *discovery.EndpointSlice) {
 	}
 }
 
-// dropTopologyOnV1 on V1 request wipes the DeprecatedTopology field  and copies
-// the NodeName value into DeprecatedTopology
+// dropTopologyOnV1 on V1 request wipes the DeprecatedTopology field.
 func dropTopologyOnV1(ctx context.Context, oldEPS, newEPS *discovery.EndpointSlice) {
 	if info, ok := genericapirequest.RequestInfoFrom(ctx); ok {
 
@@ -167,14 +165,6 @@ func dropTopologyOnV1(ctx context.Context, oldEPS, newEPS *discovery.EndpointSli
 					//Silently clear out DeprecatedTopology
 					ep.DeprecatedTopology = nil
 				}
-
-				// Set topology[hostname] to match NodeName for compatibility with old clients
-				if ep.NodeName != nil {
-					if ep.DeprecatedTopology == nil {
-						ep.DeprecatedTopology = make(map[string]string, 1)
-					}
-					ep.DeprecatedTopology[corev1.LabelHostname] = *ep.NodeName
-				}
 			}
 		}
 	}
```

Comment (swetharepakula) [reply pkg/apis/discovery/v1beta1/conversion.go:52]: If in v1beta1, `nodeName` is not populated, should we sync that with `topology[hostname]`?  Otherwise  v1beta1 -> v1, only deprecatedTopology has the hostname.

Comment (thockin) [reply pkg/apis/discovery/v1beta1/conversion.go:52]: since nodeName is a field in v1beta1, I think it's easier to just leave it mapped to the v1 field.  read-clients can fall-back on topology, but when v1beta1 goes away, write-clients need to update to the field anyway.

Comment (swetharepakula) [reply pkg/apis/discovery/v1beta1/conversion.go:63]: The only other consequence I see of this approach is that if in v1beta1 only `NodeName` is set, after the round trip, `topology[hostname]` will also be set. I don't think that is too much of an issue and it is behavior that will go away once v1beta1 is removed.

File: pkg/apis/discovery/validation/validation.go (hunk 11 above)
Comment (swetharepakula) [reply pkg/apis/discovery/validation/validation.go:120]: I had a question about it initially and just forgot to add it back in once I resolved it.

With this branch, when v1b1 controller set region it came through in v1 deprecatedTopology. I didn't investigate too much further. What is PEBKAC?

File: pkg/apis/discovery/install/install.go (hunk 13 above)
Comment (swetharepakula) [reply pkg/apis/discovery/install/install.go]: Fixed so etcd storage writes to v1beta1.

File: test/integration/etcd/data.go (hunk 14 above)
Comment (swetharepakula) [reply test/integration/etcd/data.go]: Updated to expect v1beta1 for v1.

File: pkg/controlplane/storageversionhashdata/data.go (hunk 16 above)
Comment (swetharepakula) [reply pkg/controlplane/storageversionhashdata/data.go]: changed storage version hash back to v1beta1

File: pkg/registry/discovery/endpointslice/strategy.go (hunk 18 above)
Comment (swetharepakula) [reply pkg/registry/discovery/endpointslice/strategy.go]: updated so we check for v1beta1 requests and return early.

File: pkg/apis/discovery/validation/validation.go (hunk 11 above)
Comment (thockin) [reply pkg/apis/discovery/validation/validation.go:120]: PEBKAC - Problem Exists Between Keyboard And Chair

**Review by thockin — COMMENTED** [review]
YAY!

/lgtm
/approve

**Comment by aojea** [comment]
Impressive :clap:

**Comment by aojea** [comment]
```
=== RUN   TestWatch
I0306 20:06:14.051962   69532 client.go:360] parsed scheme: "endpoint"
I0306 20:06:14.052118   69532 endpoint.go:68] ccResolverWrapper: sending new addresses to cc: [{unix://localhost:65802040492353799000  <nil> 0 <nil>}]
I0306 20:06:14.059865   69532 client.go:360] parsed scheme: "endpoint"
I0306 20:06:14.060215   69532 endpoint.go:68] ccResolverWrapper: sending new addresses to cc: [{unix://localhost:65802040492353799000  <nil> 0 <nil>}]
I0306 20:06:14.066951   69532 client.go:360] parsed scheme: "endpoint"
I0306 20:06:14.067180   69532 endpoint.go:68] ccResolverWrapper: sending new addresses to cc: [{unix://localhost:65802040492353799000  <nil> 0 <nil>}]
E0306 20:06:17.870930   69532 watcher.go:218] watch chan error: etcd event received with PrevKv=nil (key="/registry/ingresses/test/foo-ingress", modRevision=3, type=DELETE)
    resttest.go:1559: unexpected result from result channel
--- FAIL: TestWatch (4.07s)
```
/retest

**Comment by liggitt** [comment]
https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/99662/pull-kubernetes-unit/1368295059291639808

https://storage.googleapis.com/k8s-gubernator/triage/index.html?date=2021-03-06&pr=1&test=TestAdmit

```
=== RUN   TestAdmit/match_dry_run_side_effects_None
W0306 20:42:50.473031   82007 mutation_detector.go:53] Mutation detector is enabled, this will result in memory leakage.
W0306 20:42:50.473220   82007 mutation_detector.go:53] Mutation detector is enabled, this will result in memory leakage.
    plugin_test.go:187: 
        	Error Trace:	plugin_test.go:187
        	Error:      	Not equal: 
        	            	expected: map[string]string{"allow/key1":"value1", "mutation.webhook.admission.k8s.io/round_0_index_0":"{\"configuration\":\"test-webhooks\",\"webhook\":\"allow\",\"mutated\":false}"}
        	            	actual  : map[string]string(nil)
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1,5 +1,2 @@
        	            	-(map[string]string) (len=2) {
        	            	- (string) (len=10) "allow/key1": (string) (len=6) "value1",
        	            	- (string) (len=49) "mutation.webhook.admission.k8s.io/round_0_index_0": (string) (len=67) "{\"configuration\":\"test-webhooks\",\"webhook\":\"allow\",\"mutated\":false}"
        	            	-}
        	            	+(map[string]string) <nil>
        	            	 
        	Test:       	TestAdmit/match_dry_run_side_effects_None
        	Messages:   	match dry run side effects None: annotations not set as expected.
    --- FAIL: TestAdmit/match_dry_run_side_effects_None (0.02s)
```

**Comment by liggitt** [comment]
(the unit test flakes don't look related at all, just capturing them for referencing in flake issues)
